game-samples/src/golang/matchmaking-service/models/games.go

```diff
@@ -0,0 +1,281 @@
+	var rows []spanner.Row
+	defer iter.Stop()
+
+	for {
+		row, err := iter.Next()
+		if err == iterator.Done {
+			break
+		}
+
+		if err != nil {
+			return rows, err
+		}
+
+		rows = append(rows, *row)
+	}
+
+	return rows, nil
+}
+
+// Get players for a game
+// We only care about the playerUUID and their stats, as this is intended to be used
+// to modify players when a game is closed
+func getGamePlayers(gameUUID string, ctx context.Context, txn *spanner.ReadWriteTransaction) ([]string, []Player, error) {
+	stmt := spanner.Statement{
+		SQL: `SELECT PlayerUUID, Stats FROM players
+				INNER JOIN (
+				SELECT pUUID FROM games g, UNNEST(g.Players) AS pUUID WHERE gameUUID=@game
+				) AS gPlayers ON gPlayers.pUUID = players.PlayerUUID;`,
+		Params: map[string]interface{}{
+			"game": gameUUID,
+		},
+	}
+
+	iter := txn.Query(ctx, stmt)
+	playerRows, err := readRows(iter)
+	if err != nil {
+		return []string{}, []Player{}, err
+	}
+
+	var playerUUIDs []string
+	var players []Player
+	for _, row := range playerRows {
+		var p Player
+
+		if err := row.ToStruct(&p); err != nil {
+			return []string{}, []Player{}, err
+		}
+		if p.Stats.IsNull() {
+			// Initialize player stats
+			p.Stats = spanner.NullJSON{Value: PlayerStats{
+				Games_played: 0,
+				Games_won:    0,
+			}, Valid: true}
+		}
+
+		players = append(players, p)
+		playerUUIDs = append(playerUUIDs, p.PlayerUUID)
+	}
+
+	return playerUUIDs, players, nil
+}
+
+// Provided a game UUID, determine the winner
+// Current implementation is a random player from the list of players assigned to the game
+func determineWinner(playerUUIDs []string) string {
+	if len(playerUUIDs) == 0 {
+		return ""
+	}
+
+	var winnerUUID string
+
+	rand.Seed(time.Now().UnixNano())
+	offset := rand.Intn(len(playerUUIDs))
+	winnerUUID = playerUUIDs[offset]
+	return winnerUUID
+}
+
+// Given a list of players and a winner's UUID, update players of a game
+// Updating players involves closing out the game (current_game = NULL) and
+// updating their game stats. Specifically, we are incrementing games_played.
+// If the player is the determined winner, then their games_won stat is incremented.
+func updateGamePlayers(players []Player, winnerUUID string, gameUUID string,
+	ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+	for _, p := range players {
+		// Modify stats
+		var pStats PlayerStats
+		json.Unmarshal([]byte(p.Stats.String()), &pStats)
+
+		pStats.Games_played = pStats.Games_played + 1
+
+		if p.PlayerUUID == winnerUUID {
+			pStats.Games_won = pStats.Games_won + 1
+		}
+		updatedStats, _ := json.Marshal(pStats)
+		p.Stats.UnmarshalJSON(updatedStats)
+
+		// Update player
+		stmt := spanner.Statement{
+			SQL: `UPDATE players SET current_game = NULL, stats=@pStats WHERE current_game=@game AND playerUUID=@player`,
+			Params: map[string]interface{}{
+				// "stats": g.Stats,
+				"game":   gameUUID,
+				"pStats": p.Stats,
+				"player": p.PlayerUUID,
+			},
```

Comment:
Here also: It might be good to consider using mutations if we are not going to read this update in the same transaction anyways.

Reply:
so this is in another transaction that does other modifications (using DML). It has a read, but the read is done at the beginning of the transaction to get the current snapshot of players' stats. If I convert the whole transaction to mutations, that would work, right? (This might be the best section to use mutations, because closing a game takes almost 1s for 100 players on my test instance)

Reply:
Yes, that should work. Mutations are buffered locally until the transaction is committed, so you cannot read any changes that are made by those mutations until the transaction is committed. But you can read other data in the transaction, and it will be guaranteed by Spanner that the data that you read are still unchanged when the transaction is committed. So I agree, this is probably a good place to showcase mutations.

Reply:
Well, anecdotal test has DML closing a game in between 600-900ms. Mutation closes in 30-100ms.